hazelcast/src/test/java/com/hazelcast/internal/cluster/impl/MembershipUpdateTest.java

```diff
@@ -810,6 +810,8 @@ public void testListenerRegistrationWhileANewMemberJoining() {
         assertClusterSizeEventually(2, hz1, hz2);
 
         IMap<Object, Object> map = hz1.getMap("test");
+        // give a time to initialize ListenerAdapters on members
+        sleepSeconds(2);
```

Comment:
i am not a big fan of adding sleep to test code.
I suppose the issue is that listener registration on hz2 is not done by the time map.put is executed, because proxy registration propagates via async events.
I think that if you obtain the map proxy from both members (so just add a `hz2.getMap("test")` statement and ignore its result), it should be enough to ensure the listener is registered on both members before proceeding to the `put` action.